### DISCUSSION TITLE
Integration Issue with Firebase and NextAuth v5 in Next.js 14.0

### DISCUSSION THREAD

**Discussion by TomBell95** [nextauthjs/next-auth#9549] [Help]
## Description:
I'm currently working on integrating Firebase authentication with NextAuth v5 in a Next.js 14.0 application, following the [Adding Authentication](https://nextjs.org/learn/dashboard-app/adding-authentication) guide. However, I'm encountering difficulties in getting Firebase to work alongside NextAuth for user authentication and session management.

## Expected Behavior:
Users should be able to log in using Firebase authentication, and their session state should be managed by NextAuth.

## Current Behavior:
I can log the user in on the server, however the client is not updated, so I when access information on the user (i.e. refreshed auth tokens), it is not available.

When i access `auth.currentUser` I get `null`

### LoginForm.tsx
```typescript
'use client'

const LoginForm = () => {
  // ...
  const handleLogin = async (event: React.FormEvent) => {
    event.preventDefault();
    try {
      const formData = new FormData();
      formData.append('email', email);
      formData.append('password', password);
      await signInAction(undefined, formData);
    } catch (error) {
      // handle error
    }
  };
  // ...
};

```

### Firebase Actions (/firebase/actions.ts):
```typescript
'use server';
export async function signInAction(prevState: string | undefined, formData: FormData) {
  try {
    await signIn('credentials', formData);
  } catch (error) {
    // handle error
  }
}
```

### NextAuth (/auth.ts):
```typescript
export const { auth, signIn, signOut } = NextAuth({
  providers: [
    Credentials({
      async authorize(credentials) {
        // ...
        if (parsedCredentials.success) {
          const { email, password } = parsedCredentials.data;
          const user = await firebaseSignIn(email, password);
          if (!user) return null;
          return user;
        }
        console.log('Invalid credentials');
        return null;
      },
    }),
  ],
  pages: {
    signIn: '/authentication/login',
  },
  callbacks: {
    authorized({ auth, request: { nextUrl } }) {
      try {
        const isLoggedIn = !!auth?.user
        const isOnAuthPages = nextUrl.pathname.startsWith('/authentication')
        if (!isOnAuthPages) {
          if (isLoggedIn) {
            return true
          }
          return false
        } else if (isLoggedIn) {
          return NextResponse.redirect(new URL('/', nextUrl))
        }
        return true
      } catch (error) {
        // handle error
      }
    },
});

```

**Comment by Adel-Alwaqfi** [comment]
If you import auth from auth.ts into your pages, you will be able to extract the email as in the following example

```typescript
import { auth as AppAuth } from '@/auth';


export default async function Page() {
  const temp = await AppAuth() 
  
    return <p>Hello {temp?.user.email}</p>
}
```

However, this doesn't solve the problem. Even though the login and signout logic is working, the user doesn't maintain authentication by the Firebase service across pages.

**Reply by TomBell95** [reply]
So yes, using that approach you can retrieve the user from `next-auth` but retrieving the firebase user - i.e. `currentUser.getIdToken` - is not possible from what I have tried/read.

**Reply by Adel-Alwaqfi** [reply]
Exactly, this is the same case with me.

**Reply by Adel-Alwaqfi** [reply]
@TomBell95 
Any luck solving the problem?

**Reply by TomBell95** [reply]
Nope... however a workaround is to store the the authToken in the session and then in the actions use `getServerSession` to extract the session token from next-auth - its a `v4` approach whilst using `v5` but I honestly do not see a way in v5 to do this. Any suggestions welcome!

##  /api/auth/[...nextauth]/token.ts
```javascript
export async function refreshAccessToken(tokenData: { refreshToken: string | undefined }) {
  if (!tokenData.refreshToken) return
  try {
    const headers: HeadersInit = new Headers({
      'Content-Type': 'application/json',
    })
    
    // call the firebase auth token service (workaround)
    const response = await fetch(
      `https://securetoken.googleapis.com/v1/token?key=${process.env.NEXT_PUBLIC_FIREBASE_API_KEY}`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify({
          grant_type: 'refresh_token',
          refresh_token: tokenData.refreshToken,
        }),
      },
    )

    const refreshedTokens = await response.json()

    if (!response.ok) {
      throw refreshedTokens
    }

    return {
      accessToken: refreshedTokens.access_token,
      accessTokenExpires: Date.now() + refreshedTokens.expires_in * 1000,
      refreshToken: refreshedTokens.refresh_token ?? tokenData.refreshToken,
    }
  } catch (error) {
    return {
      ...tokenData,
      error: 'RefreshAccessTokenError',
    }
  }
}

```

##  /actions.ts
```javascript
const fetchBlah = async (): Promise<BlahResponse> => {
  try {
    const session = await getServerSession(authOptions)
    const accessToken = session?.accessToken
    const headers: HeadersInit = new Headers({
      'Content-Type': 'application/json',
      Authorization: `Firebase ${accessToken}`,
    })
    ...
  } catch (error) {
    throw error
  }
}
```

**Reply by Adel-Alwaqfi** [reply]
so you didn't change anything in the previous implementation? you only added those two functions and that's it?

**Reply by Samaritain225** [reply]
I had the same issue in my project (nextauth v4) and the work around I used there was to use the firebase login function (sign in with email and password) like the following: ` await signInWithEmailAndPassword(auth, email, password),
        await signIn("credentials", {
          redirect: true,
          email,
          password,
          callbackUrl: "/dashboard",
        }),
      ]);
`.
I'm not very okay with the method but it allowed me to have access to the current logged in user. In my case I only used credentials provider as once the user is logged in I have to make another request to my separated API using the token to get some specific info.  I hope someone found something much better. Currently I'm facing the same issue with the v5, I guess I will have to use the same method again until a better way is found. Hope this help

**Reply by Adel-Alwaqfi** [reply]
@Samaritain225 
I changed the whole thing by moving into using the firebase-admin SDK in the server action. This way I don't need to have the logged user instance. The only thing you need it to setup the firebase-admin once and that's it.

**Reply by shadiramadan** [reply]
@TomBell95 For `await firebaseSignIn(email, password);` are you using firebase `signInWithEmailAndPassword` or your own API calls?

**Reply by Samaritain225** [reply]
@Adel-Alwaqfi can you give more detail on how you proceed? I would really appreciate

**Reply by Samaritain225** [reply]
@shadiramadan I use : signInWithEmailAndPassword

**Reply by ionutcirja** [reply]
Where should I call firebase signOut? Thanks
